src/components/TextInput/PasswordInput.js

```diff
@@ -0,0 +1,180 @@
+import React from 'react';
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import Icon from '../Icon';
+import { textInputProps, togglePasswordVisibilityIconProps } from './util';
+
+export default class PasswordInput extends React.Component {
+  state = {
+    type: 'password',
+  };
+
+  togglePasswordVisibility = () => {
+    this.setState({
+      type: this.state.type === 'password' ? 'text' : 'password',
+    });
+  };
+
+  render() {
+    const {
+      alt,
+      labelText,
+      className,
+      id,
+      placeholder,
+      onChange,
+      onClick,
+      hideLabel,
+      invalid,
+      invalidText,
+      helperText,
+      light,
+      ...other
+    } = this.props;
+    const errorId = id + '-error-msg';
+    const textInputClasses = classNames('bx--text-input', className, {
+      'bx--text-input--light': light,
+    });
+    const labelClasses = classNames('bx--label', {
+      'bx--visually-hidden': hideLabel,
+    });
+    const label = labelText ? (
+      <label htmlFor={id} className={labelClasses}>
+        {labelText}
+      </label>
+    ) : null;
+    const sharedTextInputProps = {
+      id,
+      onChange: evt => {
+        if (!other.disabled) {
+          onChange(evt);
+        }
+      },
+      onClick: evt => {
+        if (!other.disabled) {
+          onClick(evt);
+        }
+      },
+      placeholder,
+      type: this.state.type,
+      className: textInputClasses,
+      ...other,
+    };
+    const error = invalid ? (
+      <div className="bx--form-requirement" id={errorId}>
+        {invalidText}
+      </div>
+    ) : null;
+    const passwordIsVisible = this.state.type === 'text';
+    const input = (
+      <>
+        <input
+          {...textInputProps({ invalid, sharedTextInputProps, errorId })}
+          data-toggle-password-visibility={this.state.type === 'password'}
+        />
+        <button
+          className="bx--text-input--password__visibility bx--tooltip__trigger bx--tooltip--icon__bottom"
+          aria-label={alt || `${passwordIsVisible ? 'Hide' : 'Show'} password`}
+          onClick={this.togglePasswordVisibility}>
+          <Icon
+            {...togglePasswordVisibilityIconProps({
+              passwordIsVisible,
+              alt,
+            })}
+          />
+        </button>
+      </>
+    );
+    const helper = helperText ? (
+      <div className="bx--form__helper-text">{helperText}</div>
+    ) : null;
+
+    return (
+      <div className="bx--form-item bx--text-input-wrapper">
```

Comment:
Can we render the stateless version here...?

Reply:
did you mean render a ControlledPassword Input here?

Reply:
Yep!